### PR TITLE
add a dep-compare python script

### PR DESCRIPTION
**Pull request by garbageslam**
At current revision it prints things like this

```
➜  mobilecoin git:(dep-compare-tool) ./tools/dep-compare.py
aead: root = {'0.4.1'}, fog_ledger = {'0.4.2'}
aho-corasick: root = {'0.7.10'}, consensus = {'0.7.8'}
aho-corasick: root = {'0.7.10'}, fog_ingest = {'0.7.6'}
aho-corasick: root = {'0.7.10'}, fog_ledger = {'0.7.18'}
aho-corasick: root = {'0.7.10'}, fog_view = {'0.7.6'}
anyhow: root = {'1.0.52'}, consensus = {'1.0.26'}
anyhow: root = {'1.0.52'}, fog_ingest = {'1.0.26'}
anyhow: root = {'1.0.52'}, fog_ledger = {'1.0.26'}
anyhow: root = {'1.0.52'}, fog_view = {'1.0.26'}
atty: root = {'0.2.14'}, fog_ingest = {'0.2.13'}
autocfg: root = {'1.0.0'}, consensus = {'1.0.1'}
autocfg: root = {'1.0.0'}, fog_ingest = {'1.0.1'}
autocfg: root = {'1.0.0'}, fog_ledger = {'1.0.1'}
autocfg: root = {'1.0.0'}, fog_view = {'1.0.1'}
cmake: root = {'0.1.45'}, consensus = {'0.1.43'}
cmake: root = {'0.1.45'}, fog_ingest = {'0.1.43'}
cmake: root = {'0.1.45'}, fog_ledger = {'0.1.43'}
cmake: root = {'0.1.45'}, fog_view = {'0.1.43'}
env_logger: root = {'0.8.3'}, fog_ingest = {'0.8.4'}
env_logger: root = {'0.8.3'}, fog_ledger = {'0.8.4'}
env_logger: root = {'0.8.3'}, fog_view = {'0.8.4'}
generic-array: root = {'0.12.3', '0.14.5'}, consensus = {'0.14.4'}
generic-array: root = {'0.12.3', '0.14.5'}, fog_ingest = {'0.14.4'}
generic-array: root = {'0.12.3', '0.14.5'}, fog_ledger = {'0.14.4'}
generic-array: root = {'0.12.3', '0.14.5'}, fog_view = {'0.14.4'}
getrandom: root = {'0.2.4', '0.1.14'}, consensus = {'0.2.2'}
getrandom: root = {'0.2.4', '0.1.14'}, fog_ingest = {'0.2.2'}
getrandom: root = {'0.2.4', '0.1.14'}, fog_ledger = {'0.2.2'}
getrandom: root = {'0.2.4', '0.1.14'}, fog_view = {'0.2.3'}
half: root = {'1.5.0'}, consensus = {'1.4.1'}
half: root = {'1.5.0'}, fog_ingest = {'1.4.1'}
half: root = {'1.5.0'}, fog_ledger = {'1.4.1'}
half: root = {'1.5.0'}, fog_view = {'1.4.1'}
hermit-abi: root = {'0.1.12'}, consensus = {'0.1.6'}
hermit-abi: root = {'0.1.12'}, fog_ledger = {'0.1.6'}
hermit-abi: root = {'0.1.12'}, fog_view = {'0.1.6'}
itertools: root = {'0.10.3'}, consensus = {'0.10.1'}
itertools: root = {'0.10.3'}, fog_ingest = {'0.10.1'}
itertools: root = {'0.10.3'}, fog_ledger = {'0.10.1'}
itertools: root = {'0.10.3'}, fog_view = {'0.10.1'}
lazycell: root = {'1.2.1'}, consensus = {'1.3.0'}
lazycell: root = {'1.2.1'}, fog_ingest = {'1.3.0'}
lazycell: root = {'1.2.1'}, fog_ledger = {'1.3.0'}
lazycell: root = {'1.2.1'}, fog_view = {'1.3.0'}
libc: root = {'0.2.114'}, consensus = {'0.2.103'}
libc: root = {'0.2.114'}, fog_ingest = {'0.2.103'}
libc: root = {'0.2.114'}, fog_ledger = {'0.2.103'}
libc: root = {'0.2.114'}, fog_view = {'0.2.103'}
log: root = {'0.4.11', '0.3.9'}, consensus = {'0.4.8'}
log: root = {'0.4.11', '0.3.9'}, fog_ingest = {'0.4.8'}
log: root = {'0.4.11', '0.3.9'}, fog_ledger = {'0.4.14'}
log: root = {'0.4.11', '0.3.9'}, fog_view = {'0.4.8'}
memchr: root = {'2.4.1'}, consensus = {'2.3.0'}
memchr: root = {'2.4.1'}, fog_ingest = {'2.2.1'}
memchr: root = {'2.4.1'}, fog_ledger = {'2.4.0'}
memchr: root = {'2.4.1'}, fog_view = {'2.2.1'}
num-integer: root = {'0.1.42'}, fog_ingest = {'0.1.44'}
num-integer: root = {'0.1.42'}, fog_ledger = {'0.1.44'}
num-integer: root = {'0.1.42'}, fog_view = {'0.1.44'}
num-traits: root = {'0.2.14'}, consensus = {'0.2.11'}
ppv-lite86: root = {'0.2.10'}, consensus = {'0.2.15'}
proc-macro2: root = {'1.0.36', '0.4.30'}, consensus = {'1.0.24'}
proc-macro2: root = {'1.0.36', '0.4.30'}, fog_ingest = {'1.0.24'}
proc-macro2: root = {'1.0.36', '0.4.30'}, fog_ledger = {'1.0.27'}
proc-macro2: root = {'1.0.36', '0.4.30'}, fog_view = {'1.0.24'}
quote: root = {'1.0.15', '0.6.13'}, consensus = {'1.0.9'}
quote: root = {'1.0.15', '0.6.13'}, fog_ingest = {'1.0.9'}
quote: root = {'1.0.15', '0.6.13'}, fog_ledger = {'1.0.9'}
quote: root = {'1.0.15', '0.6.13'}, fog_view = {'1.0.9'}
rand_chacha: root = {'0.2.2', '0.3.1'}, fog_ingest = {'0.3.0'}
rand_chacha: root = {'0.2.2', '0.3.1'}, fog_ledger = {'0.3.0'}
rand_chacha: root = {'0.2.2', '0.3.1'}, fog_view = {'0.3.0'}
regex: root = {'1.3.7'}, consensus = {'1.3.4'}
regex: root = {'1.3.7'}, fog_ingest = {'1.3.1'}
regex: root = {'1.3.7'}, fog_ledger = {'1.5.4'}
regex: root = {'1.3.7'}, fog_view = {'1.3.1'}
regex-syntax: root = {'0.6.17'}, consensus = {'0.6.14'}
regex-syntax: root = {'0.6.17'}, fog_ingest = {'0.6.12'}
regex-syntax: root = {'0.6.17'}, fog_ledger = {'0.6.25'}
regex-syntax: root = {'0.6.17'}, fog_view = {'0.6.12'}
rs-libc: root = {'0.2.2'}, fog_ledger = {'0.2.3'}
same-file: root = {'1.0.6'}, fog_ingest = {'1.0.5'}
same-file: root = {'1.0.6'}, fog_ledger = {'1.0.5'}
same-file: root = {'1.0.6'}, fog_view = {'1.0.5'}
serde: root = {'1.0.136'}, consensus = {'1.0.133'}
serde: root = {'1.0.136'}, fog_view = {'1.0.133'}
serde_derive: root = {'1.0.136'}, consensus = {'1.0.133'}
serde_derive: root = {'1.0.136'}, fog_view = {'1.0.133'}
siphasher: root = {'0.3.9'}, consensus = {'0.3.7'}
siphasher: root = {'0.3.9'}, fog_ingest = {'0.3.1'}
siphasher: root = {'0.3.9'}, fog_ledger = {'0.3.1'}
siphasher: root = {'0.3.9'}, fog_view = {'0.3.1'}
smallvec: root = {'1.6.1', '0.6.13'}, consensus = {'1.2.0'}
smallvec: root = {'1.6.1', '0.6.13'}, fog_ingest = {'1.3.0'}
smallvec: root = {'1.6.1', '0.6.13'}, fog_ledger = {'1.2.0'}
smallvec: root = {'1.6.1', '0.6.13'}, fog_view = {'1.2.0'}
syn: root = {'0.15.44', '1.0.86'}, consensus = {'1.0.67'}
syn: root = {'0.15.44', '1.0.86'}, fog_ingest = {'1.0.67'}
syn: root = {'0.15.44', '1.0.86'}, fog_ledger = {'1.0.73'}
syn: root = {'0.15.44', '1.0.86'}, fog_view = {'1.0.67'}
termcolor: root = {'1.1.0'}, fog_ingest = {'1.0.5'}
termcolor: root = {'1.1.0'}, fog_ledger = {'1.0.5'}
termcolor: root = {'1.1.0'}, fog_view = {'1.0.5'}
thread_local: root = {'1.0.1'}, fog_ingest = {'0.3.6'}
thread_local: root = {'1.0.1'}, fog_view = {'0.3.6'}
unicode-normalization: root = {'0.1.17'}, consensus = {'0.1.12'}
unicode-normalization: root = {'0.1.17'}, fog_ingest = {'0.1.12'}
unicode-normalization: root = {'0.1.17'}, fog_ledger = {'0.1.12'}
unicode-normalization: root = {'0.1.17'}, fog_view = {'0.1.12'}
vcpkg: root = {'0.2.8'}, consensus = {'0.2.15'}
vcpkg: root = {'0.2.8'}, fog_ingest = {'0.2.15'}
vcpkg: root = {'0.2.8'}, fog_ledger = {'0.2.15'}
vcpkg: root = {'0.2.8'}, fog_view = {'0.2.15'}
vec_map: root = {'0.8.2'}, consensus = {'0.8.1'}
vec_map: root = {'0.8.2'}, fog_ingest = {'0.8.1'}
vec_map: root = {'0.8.2'}, fog_ledger = {'0.8.1'}
vec_map: root = {'0.8.2'}, fog_view = {'0.8.1'}
version_check: root = {'0.1.5', '0.9.3'}, consensus = {'0.9.2'}
version_check: root = {'0.1.5', '0.9.3'}, fog_ingest = {'0.9.2'}
version_check: root = {'0.1.5', '0.9.3'}, fog_ledger = {'0.9.2'}
version_check: root = {'0.1.5', '0.9.3'}, fog_view = {'0.9.2'}
winapi: root = {'0.2.8', '0.3.9'}, consensus = {'0.3.8'}
winapi: root = {'0.2.8', '0.3.9'}, fog_ingest = {'0.3.8'}
winapi: root = {'0.2.8', '0.3.9'}, fog_ledger = {'0.3.8'}
winapi: root = {'0.2.8', '0.3.9'}, fog_view = {'0.3.8'}
winapi-util: root = {'0.1.5'}, consensus = {'0.1.3'}
winapi-util: root = {'0.1.5'}, fog_ingest = {'0.1.2'}
winapi-util: root = {'0.1.5'}, fog_ledger = {'0.1.2'}
winapi-util: root = {'0.1.5'}, fog_view = {'0.1.2'}
```

I think some of these are dev dependencies, so they aren't as interesting. But I can't see an easy way to filter those out.

### CODE REVIEW

**Comment by garbageslam** [comment]
I guess this thing actually isn't that great, it doesn't keep track of when there are multiple versions of the same package

File: tools/dep-compare.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python3
+
+import toml
+
+# Given a Cargo.lock path, get a dictionary mapping package names to version numbers
+def load_cargo_lock_packages(path):
+    lock = toml.load(path)
+    return {package['name']: package['version'] for package in lock['package']}
```
Comment (remoun) [inline tools/dep-compare.py:8]: > I guess this thing actually isn't that great, it doesn't keep track of when there are multiple versions of the same package

It seems easy enough to track all versions, e.g.:

```suggestion
import toml
from collections import defaultdict

# Given a Cargo.lock path, get a dictionary mapping package names to all version numbers for that package
def load_cargo_lock_packages(path):
    lock = toml.load(path)
    result = defaultdict(list)
    for package in lock['package']:
        result[package['name']].append(package['version'])
    return result
```

**Comment by remoun** [comment]
I took the liberty of switching to sets and checking for subsets, and updated the example in the PR description.

**Review by jcape — APPROVED** [review]
The script looks like it's producing a list of differing packages, which produces is a good punchlist of things to investigate.

Longer-term, I'm wondering whether we want to invest into putting this into CI. It's strictly a good thing if we can, but there are two obvious problems with doing so:
1. It will make dependabot a bit more nuanced to wrangle.
2. There may be valid reasons to have differing versions in enclaves (e.g. some enclave specific dependency), which would make a hard failure difficult to implement reliably (perhaps impossible).

**Comment by remoun** [comment]
> Longer-term, I'm wondering whether we want to invest into putting this into CI. It's strictly a good thing if we can, but there are two obvious problems with doing so:

I'm in favour of putting this in CI!

> 1. It will make dependabot a bit more nuanced to wrangle.

I'm not sure what you mean.

> 2. There may be valid reasons to have differing versions in enclaves (e.g. some enclave specific dependency), which would make a hard failure difficult to implement reliably (perhaps impossible).

Since it's a custom script, we can add exceptions as we see fit.

File: tools/dep-compare.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    result = defaultdict(set)
+    for package in lock['package']:
+        result[package['name']].add(package['version'])
+    return result
+
+root = load_cargo_lock_packages("Cargo.lock")
+consensus = load_cargo_lock_packages("consensus/enclave/trusted/Cargo.lock")
+fog_ingest = load_cargo_lock_packages("fog/ingest/enclave/trusted/Cargo.lock")
+fog_ledger = load_cargo_lock_packages("fog/ledger/enclave/trusted/Cargo.lock")
+fog_view = load_cargo_lock_packages("fog/view/enclave/trusted/Cargo.lock")
```
Comment (iamalwaysuncomfortable) [inline tools/dep-compare.py:18]: Sounds like we want this script to do several things, and this should go into a function which automatically searches out enclave folders and does conditional checks against the deps (which the script runner could specify at the time)

**Review by iamalwaysuncomfortable — DISMISSED** [review]
I assume we're looking to check all the Cargo.lock files in the repo against the master or only enclave related ones? I second the comment that when there's proper reasons to have different versions of crates it's unlikely for neither all the engineers nor this script to know that. That means this script would need to be updated constantly, which it in all likelihood would get missed & frequently break CI. For that reason I'd be in favor for NOT having this in CI.

All in all what's the core motivation behind this script?

I think the context of what you want to be checking for changes throughout time, so it seems what would make it more useful would to be able to add command line options for checking specific conditions that are contextually interesting/necessary at the time

**Review by iamalwaysuncomfortable — APPROVED** [review]

**Pull request merged**